### PR TITLE
[WFLY-4882] Differentiate empty and undefined maximum sets.

### PR DESCRIPTION
**Pull request by sguilhen**
Depends on a WildFly Core component upgrade that includes https://github.com/wildfly/wildfly-core/pull/1477

### CODE REVIEW

**Comment by bstansberry** [comment]
retest this please

**Comment by bstansberry** [comment]
This is an API change, so it needs a version bump and a transformer to make it work properly with legacy slaves. I'll send up a link in a while showing what needs to be done.

File: security-manager/src/main/java/org/wildfly/extension/security/manager/DeploymentPermissionsResourceDefinition.java [hunk 1]
```diff
@@ -79,16 +80,24 @@
...
+    private static final ModelNode DEFAULT_MAXIMUM_SET;
+    static {
+        DEFAULT_MAXIMUM_SET = new ModelNode();
+        DEFAULT_MAXIMUM_SET.add(new ModelNode().set(PERMISSION_CLASS, "java.security.AllPermission"));
+    };
+
     static final AttributeDefinition MAXIMUM_PERMISSIONS =
             ObjectListAttributeDefinition.Builder.of(Constants.MAXIMUM_PERMISSIONS, PERMISSIONS_VALUE_TYPE)
                     .setAllowNull(true)
+                    .setDefaultValue(DEFAULT_MAXIMUM_SET)
```
Comment (stuartwdouglas) [inline security-manager/src/main/java/org/wildfly/extension/security/manager/DeploymentPermissionsResourceDefinition.java]: This requires a subsystem model version bump

**Comment by sguilhen** [comment]
Brian, a couple of comments regarding the latest commits:

- I talked to Darran about the model version bump and he recommended going for 3.0.0 and also updating the schema version to 3.0 to leave space for any changes that could be required for WF 10 (like a CVE). So skipping a schema version is intentional here but if you feel it is weird to skip a version I can rework the commit to update the schema to 2.0 (although in my opinion this should have been done when the model was bumped to 2.0.0 a while ago as IIRC a new attribute was added to allow for the specification of the jboss module in a permission).

- The maximum set transformer: In domain mode, if the maximum set is left undefined it is fine as both current and older model versions translate that to a set containing java.security.AllPermission. On the other hand, we have a problem is if we have an empty maximum set as the current model version will translate that to an empty set (thus rejecting any configured permissions) while nodes running previous versions will handle that in the same way as undefined sets (i.e. assign AllPermission and thus allow all configured permissions). These are completely opposed behaviors and we need a transformer for that.

I've decided to reject the maximum set if it is empty because of the incompatible behavior described above. An alternative would be to change the empty ModelNode to include a permission whose implies method always return false but the JDK doesn't have such permission (like a DenyAllPermission) and sending something like a RuntimePermission would be very hacky to say the minimum. So I thought rejecting the config if the maximum set is empty and documenting that in order to achieve the empty set behavior in domain mode with mixed versions one should implement a custom NoPermission (or DenyAllPermission) and configure that instead.

**Comment by sguilhen** [comment]
retest this please

**Review by bstansberry — APPROVED** [review]

**Pull request merged**